pkg/services/libraryelements/api.go

```diff
@@ -2,27 +2,44 @@ package libraryelements
 
 import (
 	"errors"
+	"fmt"
 	"net/http"
 
 	"github.com/grafana/grafana/pkg/api/response"
 	"github.com/grafana/grafana/pkg/api/routing"
-	"github.com/grafana/grafana/pkg/middleware"
+	ac "github.com/grafana/grafana/pkg/services/accesscontrol"
 	contextmodel "github.com/grafana/grafana/pkg/services/contexthandler/model"
 	"github.com/grafana/grafana/pkg/services/dashboards"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
 	"github.com/grafana/grafana/pkg/services/folder"
 	"github.com/grafana/grafana/pkg/services/libraryelements/model"
 	"github.com/grafana/grafana/pkg/web"
 )
 
 func (l *LibraryElementService) registerAPIEndpoints() {
+	authorize := ac.Middleware(l.AccessControl)
+
 	l.RouteRegister.Group("/api/library-elements", func(entities routing.RouteRegister) {
-		entities.Post("/", middleware.ReqSignedIn, routing.Wrap(l.createHandler))
-		entities.Delete("/:uid", middleware.ReqSignedIn, routing.Wrap(l.deleteHandler))
-		entities.Get("/", middleware.ReqSignedIn, routing.Wrap(l.getAllHandler))
-		entities.Get("/:uid", middleware.ReqSignedIn, routing.Wrap(l.getHandler))
-		entities.Get("/:uid/connections/", middleware.ReqSignedIn, routing.Wrap(l.getConnectionsHandler))
-		entities.Get("/name/:name", middleware.ReqSignedIn, routing.Wrap(l.getByNameHandler))
-		entities.Patch("/:uid", middleware.ReqSignedIn, routing.Wrap(l.patchHandler))
+		uidScope := ScopeLibraryPanelsProvider.GetResourceScopeUID(ac.Parameter(":uid"))
+		nameScope := ScopeLibraryPanelsProvider.GetResourceScopeName(ac.Parameter(":name"))
+
+		if l.features.IsEnabled(featuremgmt.FlagLibraryPanelRBAC) {
+			entities.Post("/", authorize(ac.EvalPermission(ActionLibraryPanelsCreate)), routing.Wrap(l.createHandler))
+			entities.Delete("/:uid", authorize(ac.EvalPermission(ActionLibraryPanelsDelete, uidScope)), routing.Wrap(l.deleteHandler))
+			entities.Get("/", authorize(ac.EvalPermission(ActionLibraryPanelsRead)), routing.Wrap(l.getAllHandler))
+			entities.Get("/:uid", authorize(ac.EvalPermission(ActionLibraryPanelsRead, uidScope)), routing.Wrap(l.getHandler))
+			entities.Get("/:uid/connections/", authorize(ac.EvalPermission(ActionLibraryPanelsRead, uidScope)), routing.Wrap(l.getConnectionsHandler))
+			entities.Get("/name/:name", authorize(ac.EvalPermission(ActionLibraryPanelsRead, nameScope)), routing.Wrap(l.getByNameHandler))
+			entities.Patch("/:uid", authorize(ac.EvalPermission(ActionLibraryPanelsWrite, uidScope)), routing.Wrap(l.patchHandler))
+		} else {
+			entities.Post("/", routing.Wrap(l.createHandler))
```

Comment:
Question: what are the current (without RBAC) safeguards for library panels? Do we require access to a folder/dashboard to CRUD a library panel?

Reply:
Existing requirements without RBAC:
To create a library panel, a user must have edit permissions on the folder that will contain the library panel
https://github.com/grafana/grafana/blob/38c3483594080b6d7e12d94f95bcacf67c3ac368/pkg/services/libraryelements/database.go#L165

To read a library panel, if I understand the SQL correctly, a user needs view permissions on the containing folder
https://github.com/grafana/grafana/blob/38c3483594080b6d7e12d94f95bcacf67c3ac368/pkg/services/libraryelements/database.go#L271-L275

Updating a library panel also requires edit permissions on the containing folder:
https://github.com/grafana/grafana/blob/38c3483594080b6d7e12d94f95bcacf67c3ac368/pkg/services/libraryelements/database.go#L499-L509

Finally, deleting a library panel also requires edit permissions on the containing folder:
https://github.com/grafana/grafana/blob/38c3483594080b6d7e12d94f95bcacf67c3ac368/pkg/services/libraryelements/database.go#L216-L218

Let me know if you have any other questions :)

Reply:
Nice, thanks for this, this is super helpful!

So I think the options are:
1) have generic permissions to CRUD library panels + require access to the folder where the library panel resides in order to edit the library panel (this is similar to what we've done with annotations);
2) scope library panel permissions to a folder, and automatically grant library panel permissions to a user when they're set to folder viewer/editor (similarly as we've done with alerting);
3) scope library panel permissions to individual library panels + require access to the folder where the library panel resides in order to edit the library panel (this is probably unnecessarily fine-grained)

All of these approaches have some benefits and drawbacks. The current implementation seems to be mostly in line with approach 2), is that correct?

Reply:
I think that's right. I more or less tried to copy how dashboard permissions seem to be handled, where they inherit permissions from the folder(s) they are contained in, but they can't have a greater level of access than the folder they're in. (e.g. can't have edit access to a library panel contained in a folder they only have view permissions for)

Reply:
Nice nice! I also think that approach makes the most sense.

> but they can't have a greater level of access than the folder they're in. (e.g. can't have edit access to a library panel contained in a folder they only have view permissions for)

This is an interesting one, I could see arguments in both directions. But I guess we can start with that, and see what feedback we get from the users.